File: apps/Voice-Biomarker/iOS/VoiceVitals/VoiceVitals/Core/AudioUtils.swift
import Foundation

/// Lightweight waveform helpers to feed the emotion model cleaner input.
///
/// Two issues hurt accuracy with real recordings (vs the model's clean, all-speech
/// training clips): (1) dead air before/after speech dilutes the model's mean-pool,
/// and (2) speech longer than the 3 s window gets cropped. These helpers trim
/// silence and split long speech into overlapping windows for logit averaging.
enum AudioUtils {

    /// Energy-based trim of leading/trailing silence (simple VAD).
    /// Keeps frames whose RMS is ≥ `relThreshold` × the loudest frame, plus a margin.
    static func trimSilence(_ x: [Float],
                            frame: Int = 320,            // 20 ms @ 16 kHz
                            relThreshold: Float = 0.08) -> [Float] {
        guard x.count > frame else { return x }

        var energies: [Float] = []
        energies.reserveCapacity(x.count / frame + 1)
        var i = 0
        while i < x.count {
            let end = min(i + frame, x.count)
            var sum: Float = 0
            for j in i..<end { sum += x[j] * x[j] }
            energies.append((sum / Float(end - i)).squareRoot())
            i += frame
        }

        guard let maxE = energies.max(), maxE > 1e-6 else { return x }
        let thr = maxE * relThreshold
        guard let firstVoiced = energies.firstIndex(where: { $0 >= thr }),
              let lastVoiced = energies.lastIndex(where: { $0 >= thr }) else { return x }

        let start = max(0, (firstVoiced - 1) * frame)            // one-frame margin
        let stop  = min(x.count, (lastVoiced + 2) * frame)
        guard stop > start else { return x }
        return Array(x[start..<stop])
    }

    /// Split into up to `maxWindows` windows of `size` samples for multi-window
    /// averaging. Short input returns a single window (caller pads/tiles it).
    static func windows(_ x: [Float], size: Int, maxWindows: Int = 3) -> [[Float]] {
        guard x.count > size else { return [x] }
        let count = min(maxWindows, Int((Double(x.count) / Double(size)).rounded(.up)))
        if count <= 1 { return [Array(x.prefix(size))] }
        let stride = (x.count - size) / (count - 1)
        return (0..<count).map { Array(x[$0 * stride ..< $0 * stride + size]) }
    }
}

File: apps/Voice-Biomarker/Android/app/src/main/java/ai/zetic/voicevitals/core/AudioUtils.kt
package ai.zetic.voicevitals.core

import kotlin.math.ceil
import kotlin.math.min
import kotlin.math.sqrt

/**
 * Lightweight waveform helpers to feed the emotion model cleaner input.
 *
 * Two issues hurt accuracy with real recordings (vs the model's clean, all-speech
 * training clips): (1) dead air before/after speech dilutes the model's mean-pool,
 * and (2) speech longer than the 3 s window gets cropped. These helpers trim
 * silence and split long speech into overlapping windows for logit averaging.
 */
object AudioUtils {

    /**
     * Energy-based trim of leading/trailing silence (simple VAD).
     * Keeps frames whose RMS is >= [relThreshold] x the loudest frame, plus a margin.
     */
    fun trimSilence(
        x: FloatArray,
        frame: Int = 320,            // 20 ms @ 16 kHz
        relThreshold: Float = 0.08f,
    ): FloatArray {
        if (x.size <= frame) return x

        val energies = ArrayList<Float>(x.size / frame + 1)
        var i = 0
        while (i < x.size) {
            val end = min(i + frame, x.size)
            var sum = 0f
            for (j in i until end) sum += x[j] * x[j]
            energies.add(sqrt(sum / (end - i)))
            i += frame
        }

        val maxE = energies.maxOrNull() ?: return x
        if (maxE <= 1e-6f) return x
        val thr = maxE * relThreshold
        val firstVoiced = energies.indexOfFirst { it >= thr }
        val lastVoiced = energies.indexOfLast { it >= thr }
        if (firstVoiced < 0 || lastVoiced < 0) return x

        val start = maxOf(0, (firstVoiced - 1) * frame)            // one-frame margin
        val stop = min(x.size, (lastVoiced + 2) * frame)
        if (stop <= start) return x
        return x.copyOfRange(start, stop)
    }

    /**
     * Split into up to [maxWindows] windows of [size] samples for multi-window
     * averaging. Short input returns a single window (caller pads/tiles it).
     */
    fun windows(x: FloatArray, size: Int, maxWindows: Int = 3): List<FloatArray> {
        if (x.size <= size) return listOf(x)
        val count = min(maxWindows, ceil(x.size.toDouble() / size.toDouble()).toInt())
        if (count <= 1) return listOf(x.copyOfRange(0, size))
        val stride = (x.size - size) / (count - 1)
        return (0 until count).map { x.copyOfRange(it * stride, it * stride + size) }
    }
}
